File: 0048-rotate-image/0048-rotate-image.go
func rotate(matrix [][]int)  {
   l,r := 0,len(matrix)-1
   for l < r {
        for i:=0; i<r-l; i++ {
            top,bottom := l,r
            topLeft := matrix[top][l+i]
            matrix[top][l+i] = matrix[bottom-i][l]
            matrix[bottom-i][l] = matrix[bottom][r-i]
            matrix[bottom][r-i] = matrix[top+i][r]
            matrix[top+i][r] = topLeft
        }
        l++; r--
   }
}

File: 0048-rotate-image/0048-rotate-image.java
class Solution {
    public void rotate(int[][] matrix) {
        int left = 0, right = matrix.length - 1;
        while(left < right){
            int top = left, bottom = right;
            for(int i=0; i<right-left; ++i){
                int topLeft = matrix[top][left+i];
                matrix[top][left+i] = matrix[bottom-i][left];
                matrix[bottom-i][left] = matrix[bottom][right-i];
                matrix[bottom][right-i] = matrix[top+i][right];
                matrix[top+i][right] = topLeft;
            }
            ++left; --right;
        }
    }
}